File: python/nndeploy/dag/graph_runner.py
# graph_runner.py

from __future__ import annotations
import json
import time
import copy
import logging
import traceback
from typing import Dict, Any, Tuple, List
import argparse

import nndeploy.base
import nndeploy.device
from .graph import Graph
from .node import Node, add_global_import_lib, import_global_import_lib
from .edge import Edge
from .base import EdgeTypeInfo
from .base import io_type_to_name

class NnDeployGraphRuntimeError(RuntimeError):
    def __init__(self, status, msg):
        self.status = status
        self.msg = msg
        super().__init__(f"Graph run failed: {status}, {msg}")

class GraphRunner:
    def __init__(self):
        self.graph = None
        self.is_cancel = False

    def _check_status(self, status):
        if status != nndeploy.base.StatusCode.Ok:
            msg = status.get_desc()
            raise NnDeployGraphRuntimeError(status, msg)

    def _build_graph(self, graph_json_str: str, name: str):
        self.graph = Graph(name)
        if self.args is not None:
            for item in self.args.node_param:
                self.graph.set_node_value(item)
        status = self.graph.deserialize(graph_json_str)
        return self.graph, status
    
    def cancel_running(self):
        if self.graph is not None:
            flag = self.graph.interrupt()
            self.is_cancel = True
            if not flag:
                raise RuntimeError(f"interrupt failed")

    def get_run_status(self):
        if self.graph is None:
            return "{}"
        run_status_map = self.graph.get_nodes_run_status_recursive()
        # print(f"run_status_map: {run_status_map}")
        json_obj = {}
        for node_name, run_status in run_status_map.items():
            status = run_status.get_status()
            if status == "INITING":
                json_obj[node_name] = {"time": -1.0, "status": status}
            elif status == "INITED":
                json_obj[node_name] = {"time": run_status.init_time, "status": status}
            else:
                json_obj[node_name] = {"time": run_status.average_time, "status": status}
        return json_obj

    def release(self):
        if self.graph is not None:
            self.graph = None
        import gc; gc.collect()
        
    def get_loop_count(self):
        if self.graph is not None:
            return self.graph.get_loop_count()
        return 0
    
    def get_parallel_type(self):
        if self.graph is not None:
            return self.graph.get_parallel_type()
        return nndeploy.base.ParallelType.Pipeline

    def run(self, graph_json_str: str, name: str, task_id: str, args: GraphRunnerArgs = None) -> Tuple[Dict[str, Any], List[Any]]:
        self.is_cancel = False
        did_deinit = False
        self.args = args
        try:           
            nndeploy.base.time_profiler_reset()
            if self.is_cancel:
                raise RuntimeError(f"graph interrupted!")

            nndeploy.base.time_point_start("deserialize_" + name)
            self.graph, status = self._build_graph(graph_json_str, name)
            self._check_status(status)
            nndeploy.base.time_point_end("deserialize_" + name)

            self.graph.set_time_profile_flag(True)
            if args is not None and args.debug:
                self.graph.set_debug_flag(True)
            # self.graph.set_parallel_type(nndeploy.base.ParallelType.Task)
            # self.graph.set_parallel_type(nndeploy.base.ParallelType.Pipeline)

            if self.is_cancel:
                raise RuntimeError(f"graph interrupted!")

            nndeploy.base.time_point_start("init_" + name)
            status = self.graph.init()
            self._check_status(status)
            nndeploy.base.time_point_end("init_" + name)

            parallel_type = self.graph.get_parallel_type()
            results = {}

            if args is not None and args.dump:
                self.graph.dump()

            if self.is_cancel:
                raise RuntimeError(f"graph interrupted!")

            nndeploy.base.time_point_start("sum_" + name)
            count = self.graph.get_loop_count()
            for i in range(count):
                t0_0 = time.perf_counter()
                status = self.graph.run()
                self._check_status(status)
                t1_0 = time.perf_counter()
                print(f"run {i} times, time: {t1_0 - t0_0}")
                if parallel_type != nndeploy.base.ParallelType.Pipeline:
                    outputs = self.graph.get_all_output()
                    for output in outputs:
                        result = output.get_graph_output()
                        if result is not None:
                            # 判断对象是否可以拷贝，如果可以拷贝，采用拷贝的方式
                            try:
                                copy_result = copy.deepcopy(result)
                            except (TypeError, AttributeError, RecursionError) as e:
                                # 如果无法深拷贝，则直接使用原对象
                                copy_result = result
                            comsumers = output.get_consumers()
                            for consumer in comsumers:
                                consumer_name = consumer.get_name()
                                io_type = io_type_to_name[consumer.get_io_type()]
                                if consumer_name not in results:
                                    results[consumer_name] = {}
                                if io_type not in results[consumer_name]:
                                    results[consumer_name][io_type] = []
                                results[consumer_name][io_type].append(copy_result)
            if parallel_type == nndeploy.base.ParallelType.Pipeline:
                for i in range(count):
                    outputs = self.graph.get_all_output()
                    results = {}
                    for output in outputs:
                        result = output.get_graph_output()
                        if result is not None:
                            # 判断对象是否可以拷贝，如果可以拷贝，采用拷贝的方式
                            try:
                                copy_result = copy.deepcopy(result)
                            except (TypeError, AttributeError, RecursionError) as e:
                                # 如果无法深拷贝，则直接使用原对象
                                copy_result = result
                            comsumers = output.get_consumers()
                            for consumer in comsumers:
                                consumer_name = consumer.get_name()
                                io_type = io_type_to_name[consumer.get_io_type()]
                                if consumer_name not in results:
                                    results[consumer_name] = {}
                                if io_type not in results[consumer_name]:
                                    results[consumer_name][io_type] = []
                                results[consumer_name][io_type].append(copy_result)
            flag = self.graph.synchronize()
            if not flag:
                raise RuntimeError(f"synchronize failed")
            nndeploy.base.time_point_end("sum_" + name)

            if self.is_cancel:
                raise RuntimeError(f"graph interrupted!")

            nodes_name = self.graph.get_nodes_name_recursive()
            
            # print(time_profiler_map)
            nndeploy.base.time_profiler_print(name)
            if count > 10:   
                nndeploy.base.time_profiler_print_remove_warmup(name, 10)

            # 另一个线程启动的函数
            # run_status_map = self.get_run_status()
            # print(run_status_map)

            if self.is_cancel:
                raise RuntimeError(f"graph interrupted!")

            nndeploy.base.time_point_start("deinit_" + name)
            status = self.graph.deinit()
            self._check_status(status)
            did_deinit = True

            is_release_cuda_cache = True
            if is_release_cuda_cache:
                try:
                    import torch
                    torch.cuda.empty_cache()
                except ImportError:
                    pass
            nndeploy.base.time_point_end("deinit_" + name)

            # run_status_map = self.get_run_status()
            # print(run_status_map)

            # time_profiler_map = {}
            # for node_name in nodes_name:
            #     time_profiler_map[node_name] = nndeploy.base.time_profiler_get_cost_time(node_name + " run()")
            # time_profiler_map["sum_" + name] = nndeploy.base.time_profiler_get_cost_time("sum_" + name)
            # time_profiler_map["init_" + name] = nndeploy.base.time_profiler_get_cost_time("init_" + name)
            # time_profiler_map["deserialize_" + name] = nndeploy.base.time_profiler_get_cost_time("deserialize_" + name)

            time_profiler_map = {}
            time_profiler_map["init_time"] = nndeploy.base.time_profiler_get_cost_time("init_" + name)
            time_profiler_map["run_time"] = nndeploy.base.time_profiler_get_cost_time("sum_" + name)

            # print(results)
            return time_profiler_map, results, status, status.get_desc()
        
        except NnDeployGraphRuntimeError as e:
            return {}, {}, e.status, e.msg

        finally:
            try:
                if self.graph is not None:
                    if not did_deinit:
                        try:
                            self.graph.deinit()
                        except Exception:
                            pass
                    try:
                        import torch
                        torch.cuda.empty_cache()
                    except Exception:
                        pass
            finally:
                pass

File: framework/source/nndeploy/dag/graph_runner.cc
#include "nndeploy/dag/graph_runner.h"

#include <chrono>
#include <iostream>

#include "nndeploy/base/time_profiler.h"
#include "nndeploy/device/tensor.h"

namespace nndeploy {
namespace dag {

GraphRunner::GraphRunner() {}

GraphRunner::~GraphRunner() {
  if (graph_ && graph_->getInitialized()) {
    graph_->deinit();
    graph_->setInitializedFlag(false);
    graph_.reset();
  }
}

std::shared_ptr<GraphRunnerResult> GraphRunner::run(
    const std::string& graph_json_str, const std::string& name,
    const std::string& task_id) {
  std::shared_ptr<GraphRunnerResult> result =
      std::make_shared<GraphRunnerResult>();
  result->status = base::kStatusCodeOk;

  name_ = name;
  task_id_ = task_id;

  // 重置时间分析器
  if (is_time_profile_) {
    base::timeProfilerReset();
  }

  // 反序列化图
  if (is_time_profile_) {
    NNDEPLOY_TIME_POINT_START("deserialize_" + name);
  }

  base::Status status = buildGraph(graph_json_str, name);
  if (status != base::kStatusCodeOk) {
    result->status = status;
    return result;
  }

  if (is_time_profile_) {
    NNDEPLOY_TIME_POINT_END("deserialize_" + name);
  }

  // 设置图的配置
  if (graph_) {
    graph_->setTimeProfileFlag(is_time_profile_);
    graph_->setDebugFlag(is_debug_);
    if (parallel_type_ != base::ParallelType::kParallelTypeNone) {
      graph_->setParallelType(parallel_type_);
    }
    graph_->setLoopMaxFlag(is_loop_max_flag_);
  }

  // 初始化图
  if (is_time_profile_) {
    base::timePointStart("init_" + name);
  }

  status = graph_->init();
  if (status != base::kStatusCodeOk) {
    result->status = status;
    return result;
  }

  if (is_time_profile_) {
    base::timePointEnd("init_" + name);
  }

  // 获取并行类型
  base::ParallelType parallel_type = graph_->getParallelType();

  // 转储图信息
  if (is_dump_) {
    graph_->dump();
  }

  // 运行图
  if (is_time_profile_) {
    base::timePointStart("sum_" + name);
  }

  int count = graph_->getLoopCount();
  for (int i = 0; i < count; ++i) {
    auto start_time = std::chrono::high_resolution_clock::now();

    status = graph_->run();
    if (status != base::kStatusCodeOk) {
      result->status = status;
      return result;
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(
        end_time - start_time);
    std::cout << "run " << i << " times, time: " << duration.count() / 1000.0
              << " ms" << std::endl;

    // 对于非流水线模式，立即获取输出
    if (parallel_type != base::ParallelType::kParallelTypePipeline) {
      std::vector<Edge*> outputs = graph_->getAllOutput();
      for (auto* output : outputs) {
        if (output) {
          void* ptr = output->getGraphOutputPtr();
          if (ptr == nullptr) {
            NNDEPLOY_LOGE("Graph output data ptr is nullptr\n");
            break;
          }
        }
      }
    }
  }

  // 对于流水线模式，在所有运行完成后获取输出
  if (parallel_type == base::ParallelType::kParallelTypePipeline) {
    for (int i = 0; i < count; ++i) {
      std::vector<Edge*> outputs = graph_->getAllOutput();
      std::cout << "pipeline mode" << std::endl;
      for (auto* output : outputs) {
        if (output) {
          void* ptr = output->getGraphOutputPtr();
          if (ptr == nullptr) {
            NNDEPLOY_LOGE("Graph output data ptr is nullptr\n");
            break;
          }
        }
      }
    }
  }

  // 同步
  bool sync_flag = graph_->synchronize();
  if (!sync_flag) {
    result->status = base::kStatusCodeErrorDag;
    return result;
  }

  if (is_time_profile_) {
    base::timePointEnd("sum_" + name);
  }

  // 获取性能分析结果
  if (is_time_profile_) {
    result->time_profiler_map["sum_" + name] =
        base::timeProfilerGetAverageTime("sum_" + name);
    result->time_profiler_map["init_" + name] =
        base::timeProfilerGetAverageTime("init_" + name);
    result->time_profiler_map["deserialize_" + name] =
        base::timeProfilerGetAverageTime("deserialize_" + name);

    // 打印性能分析结果
    NNDEPLOY_TIME_PROFILER_PRINT(name);
    if (count > 10) {
      NNDEPLOY_TIME_PROFILER_PRINT_REMOVE_WARMUP(name, 10);
    }
  }

  // 获取运行状态
  result->run_status_map = graph_->getNodesRunStatusRecursive();
  for (const auto& pair : result->run_status_map) {
    std::cout << pair.first << ": " << pair.second->getStatus() << std::endl;
  }

  // 反初始化图
  // graph_->deinit();

  return result;
}

void GraphRunner::set_json_file(bool is_json_file) {
  is_json_file_ = is_json_file;
}

void GraphRunner::set_dump(bool is_dump) { is_dump_ = is_dump; }

void GraphRunner::set_time_profile(bool is_time_profile) {
  // 测试手机app是否调用这个函数
  // 在/data/local/tmp下创建一个test_app.txt文件，并写入当前时间
  std::ofstream out("/data/local/tmp/test_app.txt");
  out << "set_time_profile: " << is_time_profile << std::endl;
  out.close();

  NNDEPLOY_LOGE("set_time_profile: %s", is_time_profile ? "true" : "false");
  is_time_profile_ = is_time_profile;
}

void GraphRunner::set_debug(bool is_debug) { is_debug_ = is_debug; }

void GraphRunner::set_parallel_type(base::ParallelType parallel_type) {
  parallel_type_ = parallel_type;
}

void GraphRunner::set_loop_max_flag(bool is_loop_max_flag) {
  is_loop_max_flag_ = is_loop_max_flag;
}

void GraphRunner::set_node_value(const std::string& node_name,
                                 const std::string& key,
                                 const std::string& value) {
  node_value_map_[node_name][key] = value;
}

void GraphRunner::set_node_value(
    std::map<std::string, std::map<std::string, std::string>> node_value_map) {
  for (const auto& node_value_item : node_value_map) {
    for (const auto& key_value_item : node_value_item.second) {
      node_value_map_[node_value_item.first][key_value_item.first] =
          key_value_item.second;
    }
  }
}

base::Status GraphRunner::buildGraph(const std::string& graph_json_str,
                                     const std::string& name) {
  graph_ = std::make_shared<Graph>(name);
  if (!graph_) {
    return base::kStatusCodeErrorOutOfMemory;
  }

  graph_->setNodeValue(node_value_map_);

  base::Status status;
  if (is_json_file_) {
    // 从文件加载
    status = graph_->loadFile(graph_json_str);
  } else {
    // 从字符串反序列化
    status = graph_->deserialize(graph_json_str);
  }

  if (status != base::kStatusCodeOk) {
    graph_.reset();
    return status;
  }

  return base::kStatusCodeOk;
}

}  // namespace dag
}  // namespace nndeploy
